src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8333,6 +8379,312 @@ private StatementSyntax ParseLocalDeclarationStatement()
+        /// Returns null if this does not look like a deconstruction-declaration after all (equal sign missing).
+        ///
+        /// The syntax is `deconstruction-variables = expression` and can appear in local declarations and for statements.
+        /// </summary>
+        private VariableDeclarationSyntax ParseDeconstructionDeclaration(bool withEquals, bool topLevel = true)
+        {
+            if (this.CurrentToken.IsVar && this.PeekToken(1).Kind == SyntaxKind.OpenParenToken)
+            {
+                // parses `var (...) = expression`
+                return ParseDeconstructionVarForm(withEquals);
+            }
+            else if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken)
+            {
+                // parses `(...) = expression` where the elements are deconstruction variables
+                return ParseDeconstructionListForm(withEquals);
+            }
+            else if (!topLevel)
+            {
+                return ParseDeconstructionTypeIdPart();
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Parses `(..., ..., ...)` where each part is deconstruction variables (such as `type id` or `var (..., ...)` or another list `(..., ..., ...)`).
+        /// </summary>
+        private VariableDeclarationSyntax ParseDeconstructionListForm(bool withEquals)
+        {
+            var list = _pool.AllocateSeparated<VariableDeclarationSyntax>();
+
+            try
+            {
+                var openParen = this.EatToken(SyntaxKind.OpenParenToken);
+
+                if (this.CurrentToken.Kind != SyntaxKind.CloseParenToken)
+                {
+                    var variable = ParseDeconstructionDeclaration(withEquals: false, topLevel: false);
+
+                    if ((object)variable != null)
+                    {
+                        list.Add(variable);
+                    }
+
+                    while (this.CurrentToken.Kind == SyntaxKind.CommaToken && (object)variable != null)
+                    {
+                        var comma = this.EatToken(SyntaxKind.CommaToken);
+                        list.AddSeparator(comma);
+
+                        variable = ParseDeconstructionDeclaration(withEquals: false, topLevel: false);
+
+                        if ((object)variable != null)
+                        {
+                            list.Add(variable);
+                        }
+                    }
+                }
+
+                var closeParen = this.EatToken(SyntaxKind.CloseParenToken);
+
+                SyntaxToken equals = null;
+                ExpressionSyntax expression = null;
+                if (withEquals)
+                {
+                    equals = this.EatToken(SyntaxKind.EqualsToken);
+                    if (equals.IsMissing)
+                    {
+                        return null;
```

Comment:
Correct. `ParseDeconstructionListForm` is only called within `ParseDeconstructionDeclaration` which is only called in the context of holding a reset point.
The equals sign after the closing paren is the distinctive sign of a deconstruction-declaration, so if we expect one but can't find it, then it was not a deconstruction-declaration (possible a bad tuple or deconstruction-assignment instead).